pkg/app/pipedv1/plugin/kubernetes/livestate/plugin_test.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2025 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package livestate
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/pipe-cd/pipecd/pkg/app/pipedv1/plugin/kubernetes/provider"
+	"github.com/pipe-cd/pipecd/pkg/plugin/diff"
+	"github.com/pipe-cd/pipecd/pkg/plugin/sdk"
+)
+
+func makeTestManifest(t *testing.T, yaml string) provider.Manifest {
+	t.Helper()
+	manifests, err := provider.ParseManifests(yaml)
+	require.NoError(t, err)
+	require.Len(t, manifests, 1)
+	return manifests[0]
+}
+
+func makeTestDiffChange(t *testing.T, oldYAML, newYAML string) provider.DiffListChange {
+	t.Helper()
+	old := makeTestManifest(t, oldYAML)
+	new := makeTestManifest(t, newYAML)
+
+	oldData, err := old.MarshalJSON()
+	require.NoError(t, err)
+	oldUnstructured := unstructured.Unstructured{}
+	err = oldUnstructured.UnmarshalJSON(oldData)
+	require.NoError(t, err)
+
+	newData, err := new.MarshalJSON()
+	require.NoError(t, err)
+	newUnstructured := unstructured.Unstructured{}
+	err = newUnstructured.UnmarshalJSON(newData)
+	require.NoError(t, err)
+
+	result, err := diff.DiffUnstructureds(oldUnstructured, newUnstructured, old.Key().String(),
+		diff.WithEquateEmpty(),
+		diff.WithIgnoreAddingMapKeys(),
+		diff.WithCompareNumberAndNumericString(),
+	)
+	require.NoError(t, err)
+
+	return provider.DiffListChange{
+		Old:  old,
+		New:  new,
+		Diff: result,
+	}
+}
+
+func TestCalculateSyncState(t *testing.T) {
```

Comment:
[nit] It would be nice to add some testcases to cover the spec.

- The maximum count of the changes is three.
- ConfigMap and Secret are masked on the text

Reply:
Thank you!
Added at this commit https://github.com/pipe-cd/pipecd/pull/5676/commits/2306bc914ce681816ebbc293b1016453a27897cb